src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -407,142 +407,162 @@ internal virtual ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes()
+        {
+            SourceAssemblySymbol sourceAssembly = SourceModule.ContainingSourceAssembly;
+            var builder = ArrayBuilder<Cci.ExportedType>.GetInstance();
+
+            if (!OutputKind.IsNetModule())
+            {
+                var modules = sourceAssembly.Modules;
+                for (int i = 1; i < modules.Length; i++) //NOTE: skipping modules[0]
                 {
-                    GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetNetModuleDecodedWellKnownAttributeData(), builder);
+                    GetExportedTypes(modules[i].GlobalNamespace, -1, builder);
                 }
+            }
 
-                Debug.Assert(_lazyExportedTypes.IsDefault);
+            var seenTopLevelForwardedTypes = new HashSet<NamedTypeSymbol>();
+            GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetSourceDecodedWellKnownAttributeData(), builder);
 
-                _lazyExportedTypes = builder.ToImmutableAndFree();
+            if (!OutputKind.IsNetModule())
+            {
+                GetForwardedTypes(seenTopLevelForwardedTypes, sourceAssembly.GetNetModuleDecodedWellKnownAttributeData(), builder);
+            }
 
-                if (_lazyExportedTypes.Length > 0)
+            return builder.ToImmutableAndFree();
+        }
+
+        private void ReportExportedTypeNameCollisions(ImmutableArray<Cci.ExportedType> exportedTypes, DiagnosticBag diagnostics)
+        {
+            var sourceAssembly = SourceModule.ContainingSourceAssembly;
+            var exportedNamesMap = new Dictionary<string, NamedTypeSymbol>(StringOrdinalComparer.Instance);
+
+            foreach (var exportedType in exportedTypes)
+            {
+                var type = (NamedTypeSymbol)exportedType.Type;
+
+                Debug.Assert(type.IsDefinition);
+
+                if (type.IsTopLevelType())
                 {
-                    var exportedNamesMap = new Dictionary<string, NamedTypeSymbol>(StringOrdinalComparer.Instance);
+                    string fullEmittedName = MetadataHelpers.BuildQualifiedName(
+                        ((Cci.INamespaceTypeReference)type).NamespaceName,
+                        Cci.MetadataWriter.GetMangledName(type));
 
-                    // Report name collisions.
-                    foreach (var exportedType in _lazyExportedTypes)
+                    // First check against types declared in the primary module
+                    if (ContainsTopLevelType(fullEmittedName))
                     {
-                        Debug.Assert(exportedType.IsDefinition);
-
-                        if (exportedType.IsTopLevelType())
+                        if (type.ContainingAssembly == sourceAssembly)
+                        {
+                            diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
+                                ErrorCode.ERR_ExportedTypeConflictsWithDeclaration, type, type.ContainingModule), NoLocation.Singleton));
```

Comment:
Done.

---
In reply to: [67205593](https://github.com/dotnet/roslyn/pull/12004#discussion_r67205593) [](ancestors = 67205593)